platforms/software/maven/src/main/java/org/gradle/api/publish/maven/tasks/PublishToMavenRepository.java

```diff
@@ -55,41 +52,33 @@
  */
 @DisableCachingByDefault(because = "Not worth caching")
 public abstract class PublishToMavenRepository extends AbstractPublishToMaven {
-    private final Transient.Var<DefaultMavenArtifactRepository> repository = varOf();
+    private final Transient<Property<MavenArtifactRepository>> repository = Transient.of(getObjectFactory().property(MavenArtifactRepository.class));
     private final Cached<PublishSpec> spec = Cached.of(this::computeSpec);
-    private final Property<Credentials> credentials = getProject().getObjects().property(Credentials.class);
+
+    public PublishToMavenRepository() {
+        getCredentials().convention(
+            getRepositoryProvider().flatMap(DefaultMavenArtifactRepository::getConfiguredCredentials)
+        );
+    }
 
     /**
      * The repository to publish to.
      *
-     * @return The repository to publish to
+     * For now, only instances of {@link DefaultMavenArtifactRepository} are supported.
      */
     @Internal
-    @ToBeReplacedByLazyProperty
-    public MavenArtifactRepository getRepository() {
+    @ReplacesEagerProperty
+    public Property<MavenArtifactRepository> getRepository() {
```

Comment:
❓ Should this be a `Property`? It looks like a nested type
/cc @jvandort

Reply:
Its difficult to say. We generally want `Property` values to be immutable and instead say mutable things should be `@Nested`. For that reason I say nested, but I think this is a third case. 

We have some mutable thing elsewhere, and we want to store a reference to that mutable thing. I'm not sure we have a pattern for this. 

If we make this nested, we would need a bunch of logic to (lazily) copy all of the properties of the source repo to this repo, and I'm not sure how we accomplish that. 

I'd probably choose to leave this as-is unless we can figure out another pattern

Reply:
Really what it looks like is we should have a `@Nested` ` `DefaultRepositorySpec` here, and then we wire the properties individually from the repository on the publications block to the repo spec on the publish task

Reply:
Considering this again, it might just make sense to not touch the repo and pom/descriptor in this PR. 

We should probably just revert this change and most of the others you added a similar comment on to. 

Both `@Nested` and `Property` (with a mutable value) do not properly fit the use case. 

What we really want is to have some immutable "POM spec" that we set as a property on the publish task, with the publishing plugin's job to create a provider that evaluates to that immutable spec and sets it on the task.

Until we have that, I would leave these particular properties untouched

Reply:
https://github.com/gradle/gradle/issues/30741